### PR TITLE
Refactor export client service models

### PR DESCRIPTION
**Pull request by jdharms**
#1012 Refactors the export client service's models to conform
to the other services and abstract away Mongo specific fields.

Signed-off-by: Daniel Harms <jdharms@gmail.com>

### CODE REVIEW

File: internal/pkg/db/mongo/models/registration.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	r.Filter.FromContract(c.Filter)
+
+	r.Encryption = EncryptionDetails{}
+	r.Encryption.FromContract(c.Encryption)
+
+	r.Compression = c.Compression
+	r.Enable = c.Enable
+	r.Destination = c.Destination
+
+	//return toContractId(r.ID, r.Uuid), nil
```
Comment (michaelestrin) [inline internal/pkg/db/mongo/models/registration.go]: Dependency on code in #878

File: internal/pkg/db/mongo/export.go [hunk 2]
```diff
@@ -14,51 +14,73 @@
...
 	s := mc.getSessionCopy()
 	defer s.Close()
 
-	reg.Created = db.MakeTimestamp()
-	reg.ID = bson.NewObjectId()
+	registration := &models.Registration{}
+	id, err := registration.FromContract(r)
+
+	if err != nil {
+		return id, err
```
Comment (michaelestrin) [inline internal/pkg/db/mongo/export.go]: Returning an "empty" value for id instead is consistent with how error conditions are handled elsewhere.

File: internal/pkg/db/mongo/export.go [hunk 3]
```diff
@@ -14,51 +14,73 @@
...
+	col := s.DB(mc.database.Name).C(db.ExportCollection)
+
+	registration.Created = db.MakeTimestamp()
 
 	// Add the registration
-	err := s.DB(mc.database.Name).C(db.ExportCollection).Insert(reg)
+	err = col.Insert(registration)
 	if err != nil {
-		return reg.ID, err
+		return id, err
```
Comment (michaelestrin) [inline internal/pkg/db/mongo/export.go]: Returning an "empty" value for id instead is consistent with how error conditions are handled elsewhere.  If not, this `if` statement is redundant.

File: internal/pkg/db/mongo/export.go [hunk 4]
```diff
@@ -14,51 +14,73 @@
...
+	}
+
+	mapped.Modified = db.MakeTimestamp()
 
-	err := s.DB(mc.database.Name).C(db.ExportCollection).UpdateId(reg.ID, reg)
-	if err == mgo.ErrNotFound {
-		return db.ErrNotFound
+	c := s.DB(mc.database.Name).C(db.ExportCollection)
+
+	var err error
```
Comment (michaelestrin) [inline internal/pkg/db/mongo/export.go]: Here to end can be simplified to:

```
	if mapped.ID.Valid() {
		return c.UpdateId(mapped.ID, mapped)
	} 
	return c.Update(bson.M{"uuid": mapped.Uuid}, mapped)
```

File: internal/pkg/db/mongo/export.go [hunk 5]
```diff
@@ -67,28 +89,59 @@ func (mc MongoClient) UpdateRegistration(reg export.Registration) error {
 // Get a registration by ID
 // UnexpectedError - problem getting in database
 // NotFound - no registration with the ID was found
-func (mc MongoClient) RegistrationById(id string) (export.Registration, error) {
-	if !bson.IsObjectIdHex(id) {
-		return export.Registration{}, db.ErrInvalidObjectId
+func (mc MongoClient) RegistrationById(id string) (contract.Registration, error) {
+	reg, err := mc.registrationById(id)
```
Comment (michaelestrin) [inline internal/pkg/db/mongo/export.go]: I would inline `registrationById` here and refactor `RegistrationById` as this is its only reference.

File: internal/pkg/db/mongo/export.go [hunk 6]
```diff
@@ -122,11 +175,11 @@ func (mc MongoClient) getRegistrations(q bson.M) ([]export.Registration, error)
 }
 
 // Get a single registration for the passed query
-func (mc MongoClient) getRegistration(q bson.M) (export.Registration, error) {
+func (mc MongoClient) getRegistration(q bson.M) (models.Registration, error) {
```
Comment (michaelestrin) [inline internal/pkg/db/mongo/export.go]: If you modify this function to return `models.Registration{}` explicitly when err != nil, you can simplify the calling code (`registrationById`, `registrationByName`).

File: internal/pkg/db/mongo/export.go [hunk 7]
```diff
@@ -67,28 +89,59 @@ func (mc MongoClient) UpdateRegistration(reg export.Registration) error {
...
+	return reg, nil
 }
 
 // Get a registration by name
 // UnexpectedError - problem getting in database
 // NotFound - no registration with the name was found
-func (mc MongoClient) RegistrationByName(name string) (export.Registration, error) {
-	return mc.getRegistration(bson.M{"name": name})
+func (mc MongoClient) RegistrationByName(name string) (contract.Registration, error) {
+	reg, err := mc.registrationByName(name)
```
Comment (michaelestrin) [inline internal/pkg/db/mongo/export.go]: I would inline `registrationByName` here and refactor `RegistrationByName` as this is its only reference.

File: internal/pkg/db/mongo/export.go [hunk 8]
```diff
@@ -14,51 +14,73 @@
...
 
 // ****************************** REGISTRATIONS ********************************
 
 // Return all the registrations
 // UnexpectedError - failed to retrieve registrations from the database
-func (mc MongoClient) Registrations() ([]export.Registration, error) {
-	return mc.getRegistrations(bson.M{})
+func (mc MongoClient) Registrations() ([]contract.Registration, error) {
+	var mrs []models.Registration
+	mrs, err := mc.getRegistrations(bson.M{})
```
Comment (michaelestrin) [inline internal/pkg/db/mongo/export.go]: I would inline `getRegistrations` and `mapRegistrations` and refactor `Registrations` as these are their own references.

File: internal/pkg/db/mongo/export.go (hunk 5 above)
Comment (jdharms) [reply internal/pkg/db/mongo/export.go]: I'm sort of fond of having the db querying and the conversion to the contract model separated out like this.

File: internal/pkg/db/mongo/export.go (hunk 6 above)
Comment (brandonforster) [reply internal/pkg/db/mongo/export.go]: > If you modify this function to return models.Registration{} explicitly when err != nil, 

Do you mean when `err == nil`?

**Comment by jdharms** [comment]
Most feedback addressed.  I left the functions unrolled for matters of taste, but can inline them if that's the consensus.

File: internal/pkg/db/mongo/export.go (hunk 6 above)
Comment (jdharms) [reply internal/pkg/db/mongo/export.go]: No, I think he means return the empty struct when there's an error (err != nil).

**Pull request merged**